### ISSUE TITLE
ReactiveSecurityContextHolder is broken

### ISSUE THREAD

**Issue by dave-fl** [spring-projects/spring-security#5690]
### Summary

`ReactiveSecurityContextHolder` is broken.  It does not always provide results and sometimes just fires the onComplete signal.

### Actual Behavior

Executes onComplete()

### Expected Behavior

Should execute onNext()

### Version

5.0.7 Release
### Sample

	@Test
	public void testTokenProvider() {
		Authentication authentication = new PreAuthenticatedAuthenticationToken("TEST", "");
		Mono<String> working = ReactiveSecurityContextHolder.getContext()
				.map(securityContext -> (String)securityContext.getAuthentication().getPrincipal());

		Mono<String> stringMono = working.subscriberContext(ReactiveSecurityContextHolder.withAuthentication(authentication));
		StepVerifier.create(stringMono).expectNext("TEST").verifyComplete();
	}

	@Test
	public void testBrokerProvider() {
		Authentication authentication = new PreAuthenticatedAuthenticationToken("TEST", "");
		Mono<String> working = ReactiveSecurityContextHolder.getContext()
				.map(securityContext -> (String)securityContext.getAuthentication().getPrincipal());
		Mono<String> broken = Mono.fromFuture(working.toFuture());
		Mono<String> stringMono = broken.subscriberContext(ReactiveSecurityContextHolder.withAuthentication(authentication));
		StepVerifier.create(stringMono).expectNext("TEST").verifyComplete();
	}

**Comment by rwinch** [comment]
Thanks for the report. If it is just an arbitrary Future I would not expect it to work. However, if the Future was produced from a `Mono` I would think it could work. If it were to work it would be a Project Reactor change.

What are your thoughts @simonbasle and @smaldini?

**Comment by dave-fl** [comment]
@rwinch In my test, the Future is produced from a `Mono`

**Comment by rwinch** [comment]
@dave-fl Right. This is why I didn't immediately close the issue since this seems somewhat reasonable to support to me.

In any case, this is something we will need Project Reactor to support. which is why I pinged Simon and Stephane. We will wait to see what their thoughts are.

**Comment by dave-fl** [comment]
Another simple example.

	@Test
	public void simple() {
		Authentication authentication = new PreAuthenticatedAuthenticationToken("SIMPLE SIMON", "");
		Mono<String> context = ReactiveSecurityContextHolder.getContext().flatMap(securityContext -> {
			return Mono.just((String)securityContext.getAuthentication().getPrincipal());
		});

		Mono<String> wrappedWithSubscribe = Mono.create(monoSink -> context.subscribe(s -> monoSink.success(s), e -> monoSink.error(e), () -> monoSink.success()));
		Mono<String> addedContext = wrappedWithSubscribe
				.subscriberContext(ReactiveSecurityContextHolder.withAuthentication(authentication));

		StepVerifier.create(addedContext).expectNext("SIMPLE SIMON").verifyComplete();
	}

**Comment by simonbasle** [comment]
@dave-fl you are switching to a `Future` and then back to a `Mono`. The `Context` is purely a Reactor concept, and is propagated along Reactive Streams signals by `Flux` and `Mono` **as long as you don't break the chain of operators** (your second example with a `subscribe` breaks that assumption).

I gather you did that first test to simulate using an external `Future`-based API in the middle?

The `MonoCompletionStage` can probably be made aware of the `Context` of a `MonoToFuture` instance, but that would only help if you're guaranteed that external API doesn't wrap the `Mono`-produced `Future`, or change it in any way.

**Comment by dave-fl** [comment]
@simonbasle Yes using Caffeine Async Cache - which uses futures.  I don't believe it changes the future as its just storing what its being provided.

**Comment by dave-fl** [comment]
Is there anything that can be done with the subscription to preserve the context?

**Comment by simonbasle** [comment]
> Is there anything that can be done with the subscription to preserve the context?

How do you mean? `Context` is accessed by an operator's internal `Subscriber` by looking at its immediate downstream `Subscriber`. If there is a stopgap in the reactive chain and there's no such `Subscriber`, then the `Context` cannot be accessed.

**Comment by dave-fl** [comment]
@simonbasle I guess what I am asking you is that if one wanted to introduce their own Mono as I have done in this example, is there not a means to patch in that context so that chain is not broken.

**Comment by simonbasle** [comment]
The sink has a `currentContext()` method so you could do `sink -> context.subscriberContext(sink.currentContext()).subscribe(...)`.

That said, subscribe inside create is an anti-pattern to me, and it achieves nothing. Probably just an attempt at reproducing the issue, but it fails by nature and not as a byproduct of a bug. (bit like saying "this code throws a NPE: `throw new NullPointerException();`" 😄).

**Comment by sepanniemi** [comment]
The same problem seems to exist if I try to use SecurityContext in RxJava2 flow. 

```
Single.fromPublisher(ReactiveSecurityContextHolder.getContext())
```

Is there any way to access the current security context if using RxJava data types in WebFlux RestController?

**Comment by simonbasle** [comment]
Nope, the reactive metadata `Context` only works as long as the reactive sequence is made of Reactor operators

**Comment by sepanniemi** [comment]
I was able to transfer the SecurityContext to RxJava flow using reactor Hooks and custom operator which copies the SecurityContext from reactor context to ThreadLocal as Single<SecurityContext> created from Mono<SecurityContext>. Needs some more testing, but looks promising for some of our existing services heavily using RxJava and soon migrating to SpringBoot2.

**Comment by ahamednijamudeensa** [comment]
I was using Webflux with AOP and unable to access ReactiveSecurityContextHolder.getContext()  inside the AOP interception as it immediately return onComplete() without returning the context. below is my sample code. is there anyway to propagate Security context into AOP interception?

`@Aspect
@Component
@Slf4j
public class NotifyAspect {

    private final NotificationServiceHelper notificationServiceHelper;

    public NotifyAspect(NotificationServiceHelper notificationServiceHelper) {
        this.notificationServiceHelper = notificationServiceHelper;
    }

    @AfterReturning(pointcut = "@annotation(com.cupid9.api.common.annotations.Notify)", returning = "returnValue")
    public void generateNotification(JoinPoint joinPoint, Object returnValue) throws Throwable {
        log.info("AfterReturning Advice - Intercepting Method : {}", joinPoint.getSignature().getName());
        
        ReactiveSecurityContextHolder.getContext().map(context -> {
            var authentication = context.getAuthentication();
            return context;
        }).subscribe()

    }

    @AfterThrowing(pointcut = "@annotation(com.cupid9.api.common.annotations.Notify)")
    public void handleNotificationException(JoinPoint joinPoint) throws Throwable {
        log.info("AfterThrowing Advice - Intercepting Method : {}", joinPoint.getSignature().getName());
        throw new InternalServerError("Error generating notification " + joinPoint.toString());
    }
}`

**Comment by lee-gilbert** [comment]
Might this also provide a solution for retrieving the SecurityContext within Spring Data AuditorAware implementations?

**Comment by rwinch** [comment]
@lee-gilbert I'm not sure. Can you provide more information?

**Comment by lee-gilbert** [comment]
@rwinch  It appears the issue is discussed without a solution here:
https://jira.spring.io/browse/DATACMNS-1231

**Comment by mgumerov** [comment]
Sorry, this might be a bit off topic but not by much because it questions if ReactiveSecurityContextHolder and Future are related at all.
So, in my project I subscribe to ReactiveSecurityContextHolder.getContext in a bean (i.e. the subscription outlives a single chain), like @Bean Mono<ConvertedPrincipal> convertedPrincipal { return ReactiveSecurityContextHolder.getContext().map(...) }
Turns out that for every new request the map() gets called again. Meaning, that desprite getContext() being Mono<>, it emits multiple values to the subscriber.

Doesn't that violate a Mono<> protocol?
And anyway, clearly that is not compatible with Future<> is it?

**Comment by simonbasle** [comment]
(sounds indeed a bit off-topic and potentially a separate issue / question, but here goes)

> emits multiple values to the subscriber - although each chain only observes one value.

these statements seem to contradict each other. are you subscribing to the `Mono` multiple times?

the `Mono` contract of _at most one `onNext` signal_ is like similar Reactive Streams contracts: it is only true in the context of a single `Subscription` - ie. for a specific `Subscriber`-`Publisher` relationship. A `Mono`/`Publisher` is like a `Subscription` factory, multiple subscriptions _can_ happen.

Bear in mind that some `Mono` will explicitly reject more than one subscription (e.g. Monos representing a transient source that can be only read once, like a unique response in a Network I/O).
Others will conceptually re-generate the source data ("cold" Monos - e.g. higher level "remote call" => perform the request again) and other will cache or just complete empty to late subscribers.

You should definitely not disconnect the `getContext` `Mono` from the request chain and make it a Bean IMHO. If you're already doing that, I assume you have implemented another kind of scope yourself, but it better be solid. You might have overlooked the potential cold and deferred nature of the `getContext` `Mono` (if someone from the @spring-projects/spring-security team can confirm it is cold).
